libgadget/fof.c

```diff
@@ -382,40 +373,96 @@ void fof_label_primary(void)
     myfree(FOF_PRIMARY_GET_PRIV(tw)->Head);
 }
 
+/* This function does the actual merging of the trees. It takes two head particles, h1 and h2,
+ * and merges 2 into h1 by setting Head[h2] = h1. It does this if and only if h1 < h2 (to avoid cycles)
+ * and it does it under a lock for h2, so that h2 cannot be merged elsewhere. h1 is not locked, to avoid deadlocking.
+ * It can thus cease to be the head of the tree, but this is fine, because it will never be merged into h2.
+ * If h1 > h2, the code swaps them and tries again.
+ * returns: 1 if a merge happened, 0 if it did not. No locks are held after return.*/
+static int
+try_merge_trees(int * h1, int * h2, MyIDType * subminID, MyIDType * subminIDTask, int * Head)
+{
+    /* We are about to change the HEAD of h2, so lock it.
+     * We need to make sure that h2 is really the head of the subtree,
+     * or not all particles will be linked.
+     * If this is not the case, unlock the particle, walk the tree again,
+     * and loop until we can get a lock.*/
+    while(1) {
+        /*Stop if we encountered h1 along the head of this tree: means this is already merged.*/
+        if(*h2 == -1)
+            return 0;
+        lock_particle(*h2);
+        if(Head[*h2] != *h2) {
+            unlock_particle(*h2);
+            *h2 = HEADa(*h1, *h2, Head);
+        }
+        else
+            break;
+    }
+    /*Get the current subtree properties: this is fine because we are under the lock.*/
+    *subminID = HaloLabel[*h2].MinID;
+    *subminIDTask = HaloLabel[*h2].MinIDTask;
+
+    /*Merge the subtree*/
+    if(*h2 > *h1) {
+        #pragma omp atomic write
```

Comment:
This write seems to be atomic anyways. Why the atomic write? Do you need the flush implied by it?

Reply:
I wasn't sure I needed it, but I wanted any threads concurrently inside of HEADa() to get to h1 instead of h2. I was worried that if a read is in progress the other threads may pick up half of h2 and half of h1.